drivers/SmartThings/jbl/src/init.lua

```diff
@@ -0,0 +1,221 @@
+  log.info("create_sse : dni = " .. tostring(device.device_network_id))
+  local conn_info = device:get_field(fields.CONN_INFO)
+
+  if not driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    log.error("create_sse : invalid connection")
+    return
+  end
+
+  local sse_url = driver.device_manager.get_sse_url(driver, device, conn_info)
+  if not sse_url then
+    log.error("failed to get sse_url")
+  else
+    log.trace("Creating SSE EventSource for " .. device.device_network_id .. ", sse_url = " .. sse_url)
+    local eventsource = EventSource.new(sse_url, {[CREDENTIAL_KEY_HEADER] = credential}, nil)
+
+    eventsource.onmessage = function(msg)
+      if msg then
+        handle_sse_event(driver, device, msg)
+      end
+    end
+
+    eventsource.onerror = function(msg)
+      log.error("Eventsource error: dni = " .. tostring(device.device_network_id) .. ", Error=" .. tostring(msg))
+      device:offline()
+    end
+
+    eventsource.onopen = function(msg)
+      log.info("Eventsource open: dni = " .. tostring(device.device_network_id))
+      device:online()
+    end
+
+    local old_eventsource = device:get_field(fields.EVENT_SOURCE)
+    if old_eventsource then
+      log.info("Eventsource Close: dni = " .. tostring(device.device_network_id))
+      old_eventsource:close()
+    end
+    device:set_field(fields.EVENT_SOURCE, eventsource)
+  end
+end
+
+local function update_connection(driver, device, device_ip, device_info)
+  local device_dni = device.device_network_id
+  log.info("update connection, dni = " .. tostring(device_dni))
+
+  local conn_info = driver.discovery_helper.get_connection_info(driver, device_dni, device_ip, device_info)
+
+  local credential = device:get_field(fields.CREDENTIAL)
+
+  conn_info:add_header(CREDENTIAL_KEY_HEADER, credential)
+
+  if driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:set_field(fields.CONN_INFO, conn_info)
+
+    create_sse(driver,device, credential)
+  end
+end
+
+
+local function find_new_connetion(driver, device)
+  log.info("find new connection for dni=" .. tostring(device.device_network_id))
+  local ip_table = discovery.find_ip_table(driver)
+  local ip = ip_table[device.device_network_id]
+  if ip then
+    device:set_field(fields.DEVICE_IPV4, ip, {persist = true})
+    local device_info = device:get_field(fields.DEVICE_INFO)
+    update_connection(driver, device, ip, device_info)
+  end
+end
+
+local function check_and_update_connection(driver, device)
+  local conn_info = device:get_field(fields.CONN_INFO)
+  if not driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:offline()
+    find_new_connetion(driver, device)
+    conn_info = device:get_field(fields.CONN_INFO)
+  end
+
+  if driver.device_manager.is_valid_connection(driver, device, conn_info) then
+    device:online()
+  end
+end
+
+local function create_monitoring_thread(driver, device, device_info)
```

Comment:
You'll see better performance and responsiveness if you can create a single task for monitoring for connection changes and fanning those changes out, instead of creating a thread-per-device. The hub environments have a single physical thread and background tasks are scheduled using cooperative multitasking as opposed to pre-emptive multitasking. So there is non-trivial latency and memory overhead when using a lot of background tasks.

Reply:
function removed